Improve @mention handling

phillipthelen opened the pull request:
This PR adds a new API call `/members/find/STRING`, that searches for users whose username contains the searched string. It also implements this autocompleting for the group chats.

If a user wants to, they can also opt out of appearing in these searches through a new settings in the preferences.


NOTE: This has to be merged/deployed together or after #10844

paglias committed:
rewrite mongoose migration to avoid using recursion

paglias committed:
fixes

paglias committed:
select more fields

paglias committed:
use lean and .update

paglias committed:
move the update username route to v3 (#10836)

SabreCat committed:
fix(tests): correct expects

SabreCat committed:
fix(tests): linting & more expects
Also one more tweak for invite validation responsiveness

SabreCat committed:
chore(news): Bailey

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.70.0

SabreCat committed:
fix(chat): less intrusive highlight and better margins

SabreCat committed:
fix(chat): more width tweakage

phillipthelen committed:
Add API Call to retrieve auto-complete options for usernames

phillipthelen committed:
Create links to users profile in chat messages

SabreCat committed:
feat(content): Oddballs Bundle
Also includes one more tweak to @mention text highlighting

SabreCat committed:
chore(sprites): compile

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.71.0

natfarleydev committed:
groupChatReceived webhook fix (#10802)

* Moved sendGroupChatReceivedWebhooks to group.sendChat function.

* Added test for new functionality.

ianoxley committed:
Set width on .custom-control-label (#10840)

Set `width: 100%` on the `.custom-control-label`.

Although `overflow-wrap: break-word` is set on the parent `.checklist-item` element, it doesn't seem to take effect unless a width is set on the label.

Dimini committed:
Very large Guild member counts overflow the badge #10753 (#10812)

greenkeeper[bot] committed:
Update superagent to the latest version 🚀 (#10848)

* fix(package): update superagent to version 4.0.0

* chore(package): update lockfile package-lock.json

paglias committed:
fix(chat): prevent duplicate messages, closes #10823

paglias committed:
Fix for #10814, prevent ParallelSave errors (#10852)

* fix(group leave): prevent ParallelSave errors while leaving a group with multiple group or challenge tasks

* fix typo

negue committed:
move computed-props to methods - refactor mountItem to use the states inside (#10853)

SabreCat committed:
feat(content): Frost Hatching Potions

SabreCat committed:
chore(sprites): compile

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.72.0

paglias committed:
fix(stable): remove progress number from petItem

Alys committed:
add two slurs - TRIGGER / CONTENT WARNING: assault, slurs, swearwords, etc

negue committed:
more checks on the item.klass, also added the specialClass checks (#10859)

SabreCat committed:
feat(content): Turkey Day 2018

SabreCat committed:
chore(sprites): compile

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.73.0

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.73.1

paglias committed:
feat(footer): always show expanded footer (#10862)

natez56 committed:
Fixes issue #10857 ("Tags have extra space at the bottom when they should be centered") (#10861)

* Fix for #10857 centered category tag text

* Fixes #10857 and #10856 display tag markdown.

phillipthelen committed:
Attach client to chat messages (#10845)

* Attach client to chat messages

* Word

* Design tweaks

* Fix potential error

phillipthelen committed:
Begin adding server-side autocomplete to web client

SabreCat committed:
chore(event): end Thanksgiving tweaks

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
4.73.2

phillipthelen committed:
Add Test to opt out of username being searchable

phillipthelen committed:
Fix issue with username highlighting

phillipthelen committed:
Correctly update message text when using autocomplete

phillipthelen committed:
remove old autocomplete component

phillipthelen committed:
Improve chat input design

phillipthelen committed:
Improve chat input design

phillipthelen committed:
Fix test errors

phillipthelen committed:
Merge branch 'develop' into phillip/autocomplete-username

phillipthelen committed:
Merge branch 'develop' into phillip/autocomplete-username

paglias commented on website/client/components/chat/chatCard.vue:
```diff
@@ -248,6 +248,16 @@ export default {
+      const link = links[i];
+      links[i].onclick = () => {
+        event.preventDefault();
+        this.$router.push({ path: link.getAttribute('href')});
```
just curious, is this necessary? i mean keeping a normal `a` tag with no `onclick` handler doesn't work?

phillipthelen replied:
No, since the markdown parser creates the `a` tags so that they open in a new tab.

paglias replied:
makes sense, thanks!

paglias replied:
Actually I just tested this and it still opens the new tab, you are missing the event in the function definition: `links[i].onclick = () => {` ->  `links[i].onclick = (event) => {`

paglias replied:
(and it should not be called `event` but `ev` or something else because `event` is defined in the global scope so it will throw an eslint error)

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -155,6 +168,30 @@
+        }
+        this.newMessage = innerText;
+      },
+      getTierIcon (member) {
```
`getTierIcon` and `tierIcon` seem the same function

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -196,20 +235,32 @@
       padding: .5em;
+      -moz-appearance: textfield-multiline;
+      -webkit-appearance: textarea;
+      background-color: -moz-field;
```
is this rule firefox specific?

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -233,6 +284,84 @@
+</style>
+
+<style lang="scss">
+  @import '~client/assets/scss/tiers.scss';
```
it looks like this file is being imported multiple times acrsso the website, resulting in duplicate css code, can you remove all the `@import '~client/assets/scss/tiers.scss';` (there are a couple more) and import it only once in `assets/scss/index`

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -233,6 +284,84 @@
+    border: 1px solid rgba(#000, 0.13);
+    background-clip: padding-box;
+    overflow: hidden;
+    -moz-transition: none;
```
it's not necessary to add browser-specific rules, we have a a plugin that automatically adds them when needed, for example in this case `transition: none;` alone will do the trick

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -233,6 +284,84 @@
+    padding: 12px 24px;
+    cursor: pointer;
+    font-size: 14px;
+    -moz-transition: none;
```
same here (about browser specific rules)

paglias commented on website/server/controllers/api-v4/members.js:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    let members = await User
+      .find({'auth.local.lowerCaseUsername': {$regex: `.*${username}.*`}, 'flags.verifiedUsername': true, 'preferences.searchableUsername': {$ne: false}})
```
let's make sure we have the necessary indexes here or this'll be super slow

paglias replied:
tbh I'm a bit worried about the number of requests this route will get and the performances, considering the $regex type of search. Can we at least make it so the prefix part is fixed so that the search is faster as explained here https://docs.mongodb.com/manual/reference/operator/query/regex/#index-use?

phillipthelen replied:
We have an index on that field.

I can also change it to work that way, however then you can only search by typing the beginning of the name correctly. But that's probably fine at least for now.

paglias replied:
I'd say it's fine for now, if people complain or there are no performance issues we can extend it but I'd prefer to start with the faster version if possible

paglias replied:
Hey @phillipthelen looks like this wasn't actually changed, the regexp

> A regular expression is a “prefix expression” if it starts with a caret (^) or a left anchor (\A), followed by a string of simple symbols. For example, the regex /^abc.*/ will be optimized by matching only against the values from the index that start with abc.

paglias replied:
Marked as unresolved

phillipthelen replied:
Sorry, I thought we'd leave the regex as is for now, since you last comment was it's fine for now. I changed it to work the way you requested

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -68,6 +76,38 @@
+        autocompleteOptions: {
+          async values (text, cb) {
+            if (text.length > 0) {
+              let suggestions = await axios.get(`/api/v4/members/find/${text}`);
```
this should be throttled (or debounced) or it'll flood the server with requests

paglias commented on website/client/components/groups/chat.vue:
```diff
@@ -1,33 +1,27 @@
                   @paste='disableMessageSendShortcut()',
-                  maxlength='3000'
+                  maxlength='3000',
+                  :contenteditable='true'
```
this will make the textarea render any type of HTML, I understand why it was added but I feel like it we should add some sort of sanitization or people will be able to copy and paste entiere word docs or html pages and have them rendered

phillipthelen replied:
Not entirely sure how to handle this (or big of an issue this actually is, since it would only affect how it's displayed in this box and not the actual chat message). I could import a library to handle this sanitizing? Writing that myself might be quite some work.

paglias replied:
I think the problem is that it would affect the displayed chat as well because markdown renders HTML code as well so it would be rendered. My biggest fear is that it could contain unsanitized JavaScript as well which could result in serious security issues but I haven't tested it

phillipthelen replied:
This wouldn't be a security issue. Since the message/input has to be sanitized on the server side to catch any security issues. The issue this could cause is that if you copy styled text, it might show up as styled text in the input box. But if unsanitized js can be posted, then that is an issue that has to be fixed on the postChat API call.

paglias reviewed (changes requested):
Left some comments, but didn't test it live. Looks like there's a small conflict to fix

paglias commented on package.json:
```diff
@@ -90,6 +90,7 @@
     "svg-url-loader": "^2.3.2",
     "svgo": "^1.0.5",
     "svgo-loader": "^2.1.0",
+    "tributejs": "^3.4.0",
```
is this used anywhere? it looks like it's installed but not used

phillipthelen replied:
It's used for the autocomplete here https://github.com/HabitRPG/habitica/pull/10872/files/8641a98107210cd1b25f06008c7ea38d41d872bb#diff-f9db085b45d481950984d26ae407e383R7

paglias replied:
Oh I see it's a peerDependency of vue-tribute

paglias commented on website/server/libs/highlightMentions.js:
```diff
@@ -0,0 +1,22 @@
+const mentionRegex = new RegExp('\\B@[-\\w]+', 'g');
+
+export async function highlightMentions (text) {
+  const mentions = text.match(mentionRegex);
```
can we move this call https://github.com/HabitRPG/habitica/commit/d934d9d759e75cc786163646f44c00e52e4cfc98#diff-ade116c537945ae89e2f5f059d4d577fR199 to a single place so we don't count mentions too as regexps can be expensive?

phillipthelen replied:
I wasn't sure how/where to handle that best. Do you have a suggestion for where to make this call?

paglias replied:
hmmm... maybe `highlightMentions()` could return both the text and the mentions? so array of [text, mentions] and use the mentions returned in api.postChat?

phillipthelen committed:
Move tier icons import to index

phillipthelen committed:
correctly name event variable

phillipthelen committed:
Debounce autocomplete calls

phillipthelen committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into autocomplete-username

# Conflicts:
#	package.json

phillipthelen committed:
optimize mention highlighting

phillipthelen commented:
Apart from the contenteditable thing, this PR should now also be ready. With that, I'm still not sure how to handle it. I guess the options are either using a library to sanitize the input here, or just leave it, since it is a local display issue that is probably not very common anyway?

phillipthelen committed:
fix failing tests

phillipthelen committed:
Fix sending private messages

paglias reviewed (changes requested):
There is a conflict and i left a comment on the regexp. Also, as discussed internally would you add the cache headers to `api.getUsernameAutocompletes`? It should be enough to add the following inside the route:

```
  res.set('Cache-Control', 'public, max-age=300000'); // 5 minutes
```

paglias commented:
Not sure if this can go directly on staging or if you want to test on delta

phillipthelen committed:
Cache username autocomplete requests

phillipthelen committed:
optimize autocomplete regex

phillipthelen committed:
Merge branch 'develop' into phillip/autocomplete-username

phillipthelen committed:
Fix lint error

phillipthelen committed:
add optional parameters to limit autocompletion to specific group

phillipthelen commented:
As discussed with @saraolson, the API now has an optional parameter to give a "context" for autocompletion, which means it can search according to a specific guild or party. For now it limits results to members of the party or guild and does not show other results.

phillipthelen committed:
Merge branch 'develop' into phillip/autocomplete-username

phillipthelen committed:
Fix non-profile urls not being usable.

phillipthelen committed:
Correctly handle autocomplete for public and private guilds

phillipthelen committed:
Add check to make sure users don’t search for parties/guilds they are not part of

phillipthelen committed:
fix lint error

phillipthelen committed:
limit autocomplete results to 5

SabreCat committed:
fix(mentioning): change default, adapt settings control to checkbox

paglias commented on website/server/controllers/api-v4/members.js:
```diff
@@ -0,0 +1,49 @@
+  url: '/members/find/:username',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    res.set('Cache-Control', 'public, max-age=300000'); // 5 minutes
```
I fee like now that we have different results based on which user makes the request this is going to create some problems

phillipthelen replied:
Oh, that's a good point. I could try if it works to only cache some of the calls? So that only the public searches are cached and all searches in private spaces are not cached. Do you know if that will work with googles caching?

paglias replied:
So we can cache everything that's the same for all users, everything else should not be cached. So I guess if the results for public searches is the same for everyone it'd be okay to cache them

phillipthelen committed:
Improve auto completing

phillipthelen commented:
This got a bunch more complicated.

What now happens: if a context+id is passed, the recent chat messages for that group are searched for matching users and those are used for autocompleting. If there are less than 5 results, it also does another search for users that have not chatted recently. It also takes into account blocked and muted users and does not autocomplete those.

paglias commented on website/server/controllers/api-v4/members.js:
```diff
@@ -23,25 +25,64 @@ api.getUsernameAutocompletes = {
+        .find(recentChatQuery)
+        .select(['profile.name', 'contributor', 'auth.local.username'])
+        .sort({'auth.timestamps.loggedin': -1})
+        .limit(5 - members.length)
```
member before this query isn't always an empty array? So members.length will always be 0, or not?

paglias reviewed (commented):
Left one comment but tbh this is going to be quite slow if at every request it has to fecth 200 messages chat messages everytime and in some cases make an extra User query

phillipthelen committed:
improve username autocomplete

phillipthelen committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into autocomplete-username

# Conflicts:
#	package.json
#	website/client/components/chat/autoComplete.vue
#	website/client/components/chat/chatCard.vue
#	website/client/components/groups/chat.vue
#	website/server/controllers/api-v3/chat.js
#	website/server/controllers/api-v3/members.js
#	website/server/controllers/api-v4/members.js

phillipthelen committed:
Fix merge issue

phillipthelen committed:
remove old code

phillipthelen commented:
I updated the title for the issue, since I removed the serverside autocomplete code. So now the PR has the @mentioning highlighting in messages and will send push notifications once I have that code done.

paglias commented on website/server/controllers/api-v3/auth.js:
```diff
@@ -308,6 +308,8 @@ api.resetPassword = {
 
       await user.save();
+    } else {
+      let user = await User.findOne({ 'auth.google.email': email }).exec();
```
is this used anywhere?

phillipthelen committed:
Send push notifications on mentions

phillipthelen committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into autocomplete-username

# Conflicts:
#	website/server/controllers/api-v3/chat.js
#	website/server/libs/pushNotifications.js
#	website/server/models/user/schema.js

phillipthelen committed:
Improve handling for sending mention notifications

phillipthelen committed:
Fix lint error

phillipthelen commented:
Implemented sending push notifications on mentioning. I made the following rules for that:

1. There are three settings. "Mention in party", "Mention in guild user is a member of", "Mention in guild user is NOT a member of"
2. If a user has party activity notifications active and is mentioned, the party activity notification is not sent. (unless user has disabled "Mention in party"
3. For private guilds, notifications are only sent if user is member of the party

I think that should cover the usecases?

paglias commented on website/client/components/chat/chatCard.vue:
```diff
@@ -217,6 +216,21 @@ export default {
+    const links = this.$refs.markdownContainer.getElementsByTagName('a');
+    for (let i = 0; i < links.length; i++) {
+      const link = links[i];
+      if (links[i].getAttribute('href').startsWith('/profile/')) {
```
what's the default action that's prevented here? I mean what's wrong with keeping the normal link? that it opens them in a new window?

phillipthelen replied:
Yes, exactly. Normally it would open in a new tab, but this way it opens in the same window as the modal. (same as when you click on their username at the top of a chat message.). This solution also still allows users to CMD+Click to open in a new tab in the background.

paglias commented on website/server/models/user/schema.js:
```diff
@@ -469,6 +469,7 @@ let schema = new Schema({
     reverseChatOrder: {$type: Boolean, default: false},
     background: String,
     displayInviteToPartyWhenPartyIs1: {$type: Boolean, default: true},
+    searchableUsername: {$type: Boolean, default: false},
```
this doesn't seem to be used anywhere

paglias reviewed (commented):
left a couple of comments and there are a few failing tests

phillipthelen committed:
Update schema.js

phillipthelen committed:
Fix failing test

phillipthelen committed:
Don't send push notification to users who aren't in the party

phillipthelen committed:
Remove tributejs from dependencies
